libxrdp/xrdp_sec.c

```diff
@@ -2298,159 +2298,190 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
     return 0;
 }
 
-/*****************************************************************************/
-/* Process a [MS-RDPBCGR] TS_UD_CS_MONITOR message.
-   reads the client monitors data */
-static int
-xrdp_sec_process_mcs_data_monitors(struct xrdp_sec *self, struct stream *s)
-{
-    int index;
-    int monitorCount;
-    int flags;
-    int x1;
-    int y1;
-    int x2;
-    int y2;
-    int got_primary;
-    struct xrdp_client_info *client_info = (struct xrdp_client_info *)NULL;
+struct display_size_description*
+process_monitor_stream(struct stream *s, int full_parameters) {
+    struct display_size_description *description = (struct display_size_description *)
+        g_malloc(sizeof(struct display_size_description), 1);
 
-    client_info = &(self->rdp_layer->client_info);
+    int NumMonitor;
+    struct monitor_info *monitor_layout;
+    struct xrdp_rect rect = {8192, 8192, -8192, -8192};
+    int got_primary = 0;
 
-    /* this is an option set in xrdp.ini */
-    if (client_info->multimon != 1) /* are multi-monitors allowed ? */
-    {
-        LOG(LOG_LEVEL_INFO, "Multi-monitor is disabled by server config");
-        return 0;
-    }
-    if (!s_check_rem_and_log(s, 8, "Parsing [MS-RDPBCGR] TS_UD_CS_MONITOR"))
-    {
-        return 1;
-    }
-    in_uint32_le(s, flags); /* flags */
-    in_uint32_le(s, monitorCount);
-    LOG_DEVEL(LOG_LEVEL_TRACE, "Received [MS-RDPBCGR] TS_UD_CS_MONITOR "
-              "flags 0x%8.8x, monitorCount %d", flags, monitorCount);
+    in_uint32_le(s, NumMonitor);
+    LOG(LOG_LEVEL_DEBUG, "  NumMonitor %d", NumMonitor);
 
-    //verify flags - must be 0x0
-    if (flags != 0)
-    {
-        LOG(LOG_LEVEL_ERROR,
-            "[MS-RDPBCGR] Protocol error: TS_UD_CS_MONITOR flags MUST be zero, "
-            "received: 0x%8.8x", flags);
-        return 1;
-    }
-    //verify monitorCount - max 16
-    if (monitorCount > 16)
+    if (NumMonitor >= CLIENT_MONITOR_DATA_MAXIMUM_MONITORS)
     {
         LOG(LOG_LEVEL_ERROR,
             "[MS-RDPBCGR] Protocol error: TS_UD_CS_MONITOR monitorCount "
-            "MUST be less than 16, received: %d", monitorCount);
-        return 1;
+            "MUST be less than 16, received: %d", NumMonitor);
+        goto exit_error;
     }
 
-    client_info->monitorCount = monitorCount;
+    description->monitorCount = NumMonitor;
 
-    x1 = 0;
-    y1 = 0;
-    x2 = 0;
-    y2 = 0;
-    got_primary = 0;
-    /* Add client_monitor_data to client_info struct, will later pass to X11rdp */
-    for (index = 0; index < monitorCount; index++)
+    for (int monitor_index = 0; monitor_index < NumMonitor; ++monitor_index)
     {
-        if (!s_check_rem_and_log(s, 20, "Parsing [MS-RDPBCGR] TS_UD_CS_MONITOR.TS_MONITOR_DEF"))
+        if (!s_check_rem_and_log(s, full_parameters == 0 ? 20 : 40, "Parsing monitor definitions."))
         {
-            return 1;
+            goto exit_error;
```

Comment:
I'm very much against using `goto` in new code, since it is prone to misuse. Even though it's usage here seems reasonable, I'd still prefer replacing it with a structured programing construct.

One option for removing the goto's is to allocate the description struct in the `xrdp_sec_process_mcs_data_monitors` function, since you already unconditionally free the description struct there currently.

Reply:
This way of exiting a method is actually what a lot of C programmers at Microsoft use (I used to work there), and other than exiting nested loops this is the only legitimate use of goto I'm aware of.

Still, I hear you. I've refactored it so that `process_monitor_stream` accepts a `display_size_description` structure  that needs to be pre-allocated by the calling functions, and since there are only two places where it's used/allocated/destroyed unconditionally, this shouldn't be a problem.